_posts/2017-04-02-generating-sitemaps-with-apache-spark.markdown

```diff
@@ -0,0 +1,218 @@
+
+```scala
+def main(args: Array[String]): Unit = {
+  val conf = new SparkConf().setAppName(this.getClass.getName)
+  val sc = new SparkContext(conf)
+  val hc = new HiveContext(sc)
+  val data = getData(hc)
+  val sitemaps = generateSitemaps(mapData(data))
+  writeSitemaps(sitemaps)
+  writeSitemapsIndex(sitemaps)
+}
+```
+
+#### Getting Data
+
+We get data from Hive, where we define schema for all data stored in HDFS. The following example fetches a show ID and its start date/time.
+
+```scala
+def getData(hc: HiveContext): RDD[Row] = {
+  hc.sql("SELECT id, start_at FROM shows")
+}
+```
+
+#### Mapping Data
+
+We partition sitemaps by date (mostly by year). This helps keep track of how many URLs Google has indexed and makes problems easier to diagnose.
+
+We start by bucketing data for simple objects into tuples of `(date, Iterable[String])`. In the following example all shows in 2016 will be grouped together.
+
+```scala
+def mapData(rows: RDD[Row]): RDD[(String, Iterable[String])] = {
+  rows.map { row =>
+    val id = row.getString(0)
+    val start_at = new DateTime(row.getLong(1))
+    val year = DateTimeFormat.forPattern("yyyy").print(start_at)
+    (date, slug)
+  }.groupByKey()
+}
+```
+
+#### Generating Sitemaps
+
+For each bucket we generate a sitemap. Each sitemap entry is a URL with a `loc`.
+
+```xml
+def getSitemapEntry(id: String): Elem = {
+  val loc = s"https://www.artsy.net/show/${id}"
+  <url><loc>{ loc }</loc></url>
+}
+```
+
+Sitemaps cannot contain more than 50,000 items according to the sitemap spec, so we generate files such as `sitemap-shows-2016-1.xml` and `sitemap-shows-2016-2.xml` when necessary.
+
+```scala
+def getSitemaps(data: RDD[(String, Iterable[String])]): RDD[(String, Elem)] = {
+  data.map {
+    case (date: String, all: Iterable[String]) =>
+      val groups = all.iterator.grouped(50000).zipWithIndex
+      groups.map {
+        case (ids: Seq[String], index: Int) =>
+          val indexLabel = if (index != 0 || groups.hasNext) s"-${index + 1}" else ""
+          val dateLabel = s"-${date}"
+          val sitemapKey = s"sitemap-shows${dateLabel}${indexLabel}"
+          val sitemapXml = <urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">{
+            ids.map(getSitemapEntry)
+          }</urlset>
+          (sitemapKey, sitemapXml)
+      }.toArray
+  }.flatMap(x => x)
+}
+```
+
+#### Writing Sitemap Files
+
+We write a file for each sitemap. The production output goes to S3.
+
+```scala
+def writeSitemaps(sitemapsXml: RDD[(String, Elem)]) = {
+  sitemapsXml.foreach {
+    case (key: String, xml: Elem) =>
+      writeFile(xml, s"${key}.xml")
+  }
+}
+```
+
+#### Writing Sitemap Index
+
+An index is a list of all the sitemaps, also written to S3.
+
+```scala
+def getSitemapsIndex(sitemapsXml: RDD[(String, Elem)]): Elem = {
+  <sitemapindex xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">{
+    sitemapsXml.collect().map {
+      case (key: String, xml: Elem) =>
```

Comment:
since the `Elem` is not used, you would usually just refer to it with the wildcard `_` in the case statement.